index.d.ts

```diff
@@ -907,7 +907,7 @@ type AppendKeyPrefix<Key, KPrefix> = KPrefix extends string
  * T function declaration *
  **************************/
 export interface TFunction<Ns extends Namespace = _DefaultNamespace, KPrefix = undefined> {
-  $TFunctionBrand: Ns;
+  $TFunctionBrand: `${$FirstNamespace<Ns>}`;
```

Comment:
btw: that might also fix the [string, string] case => ``$TFunctionBrand: `${$FirstNamespace<Ns> | _DefaultNamespace}`;``

Reply:
I get this error:

<img width="795" alt="image" src="https://github.com/i18next/i18next/assets/14237/102aa42d-1872-42a1-aae1-744b3b695ea8">

Reply:
If `Ns` is `[string, string]`, then `$FirstNamespace<Ns>` will be `string`, so `` `${string | _DefaultNamespace}` `` will also be `string` (I think).

Reply:
you're right

Reply:
the problem is in a non typed translations setup, if you're using `useTranslation('myNS')` will now return the branded TFunction with myNS

Reply:
this means you always need to use TFunction<'myNS'> also in a non typed translations setup...
hmmm....

Reply:
and also the never approach would not fix that

Reply:
This will never work with a branded TFunction:

```ts
function untypedTFunction() {
  const [t] = useTranslation('boo');
  interface Props {
    t: TFunction
  }
  const props: Props = {
    t
  }
}
```

Reply:
@pedrodurek need some expert help here...
Currently I would say we merge this PR as is or remove the branded function again.

Reply:
In a non typed setup, maybe we should avoid branding altogether or always use `never`/`string` or some consistent type for the brand?

Reply:
I think it's `$IsResourcesDefined`

Reply:
maybe this works... `{} extends CustomTypeOptions ? true : false`

Reply:
<img width="590" alt="image" src="https://github.com/i18next/i18next/assets/1086194/48fc3d91-5a37-4d13-9230-b2dbe353fa26">
<img width="1210" alt="image" src="https://github.com/i18next/i18next/assets/1086194/ba6c2308-bfae-4d31-baac-dfad5d377015">
here: https://github.com/i18next/react-i18next/blob/master/index.d.ts#L65
What do you think?

Reply:
It seems to work, what do you think of this?

```diff
diff --git a/index.d.ts b/index.d.ts
index ae1b5dd..1c40f02 100644
--- a/index.d.ts
+++ b/index.d.ts
@@ -907,7 +907,7 @@ type AppendKeyPrefix<Key, KPrefix> = KPrefix extends string
  * T function declaration *
  **************************/
 export interface TFunction<Ns extends Namespace = _DefaultNamespace, KPrefix = undefined> {
-  $TFunctionBrand: `${$FirstNamespace<Ns>}`;
+  $TFunctionBrand: 'resources' extends keyof CustomTypeOptions ? `${$FirstNamespace<Ns>}` : never;
   <
     Key extends ParseKeys<Ns, TOpt, KPrefix> | TemplateStringsArray,
     TOpt extends TOptions,
diff --git a/test/typescript/custom-types/t.test.ts b/test/typescript/custom-types/t.test.ts
index 9a8e8e6..c65c4bc 100644
--- a/test/typescript/custom-types/t.test.ts
+++ b/test/typescript/custom-types/t.test.ts
@@ -1,4 +1,4 @@
-import i18next, { TFunction } from 'i18next';
+import i18next, { CustomTypeOptions, TFunction } from 'i18next';
 
 function defaultNamespaceUsage(t: TFunction) {
   t('bar');
diff --git a/test/typescript/default-types/t.test.ts b/test/typescript/default-types/t.test.ts
index 413d7a1..ae1a78a 100644
--- a/test/typescript/default-types/t.test.ts
+++ b/test/typescript/default-types/t.test.ts
@@ -7,12 +7,15 @@ function expectErrorsForDifferentTFunctions(
   t4: TFunction<[string, string]>,
 ) {
   const fn: (t: TFunction<'plurals'>) => void = () => {};
+  const fn2: (t: TFunction) => void = () => {};
 
-  // @ts-expect-error
+  // no error - not checked when CustomTypeOptions["resources"] is not provided
   fn(t1);
-  // @ts-expect-error
   fn(t2);
-  fn(t3); // no error
-  // @ts-expect-error
+  fn(t3);
   fn(t4);
+  fn2(t1);
+  fn2(t2);
+  fn2(t3);
+  fn2(t4);
 }
```